### ISSUE TITLE
[perf] macro_beta のローカル実行が GHA の2〜4倍遅い（pytensor が C コンパイラ不在で Python フォールバック）

### ISSUE THREAD

**Issue by kome-kome** [kome-kome/financial_app#512]
親: #504（月次バッチのローカル移行）／発覚の経緯: #504 の進捗可視化（PR #511）を入れた直後のログ

## 該当

- `macro_beta_inference.py`（`python -m scripts.run_monthly --steps macro_beta`）
- ローカル Windows 実行のみ。GHA 側は無関係

## 問題

**GHA 実績 1h56m（2026-08-01）〜3h39m（2026-07-04）に対し、ローカル実行が明らかに遅い。**

2026-08-21 03:33 に起動した実行は **7時間以上走って完走せず消えた**（プロセス消滅・`macro_beta_meta` は 8/01 のまま・Egress 台帳にもエントリ無し）。パラメータは GHA と同一（`draws 800 / tune 800 / chains 2 / numpyro / adapt_diag`）。

## 観測（PR #511 のログ直結で初めて見えた）

```
WARNING (pytensor.configdefaults): g++ not detected!  PyTensor will be unable to compile
C-implementations and will default to Python. Performance may be severely degraded.
```

- `pytensor.config.cxx` は `''`（空）
- `g++` / `gcc` / `cl` のいずれも PATH に無い（`shutil.which` で確認）
- GHA の `ubuntu-latest` には g++ がある＝**この差はローカル固有**
- この警告は以前から出ていたはずだが、`Runner` の `capture_output=True` に飲まれてログに残らなかった（#504 / PR #511 で是正）

## ただし「g++ を入れれば直る」とはまだ言えない

本番実行は `--nuts-sampler numpyro` ＝ **サンプリング自体は JAX(XLA) が担う**。pytensor のグラフは JAX 関数へ変換されるため、C コンパイラ不在がどこにどれだけ効くかは未測定である。

**症状から原因を推測しない**——#500 で「Disk IO Budget 100%」を IOPS 上限と誤診し、真因（NANO の実効メモリ 408MB に DB が乗らない）に辿り着くまで遠回りした前例がある。ここでも同じ形を踏まない。

他に説明しうる要因（切り分けの対象）:

- CPU: Intel Core i5-8400（6コア・2.80GHz・2017年）。GHA ランナーはコア数こそ2だが世代が新しい
- Windows の JAX は CPU バックエンドのみ（ログにも `Unable to initialize backend 'tpu'`）
- 夜間の他プロセスとの競合（8/21 の実行中は pytest フルスイート等が並走していた）

## 改善案（測ってから入れる）

1. **現在走っている実行の所要を実測する**。PR #511 の heartbeat が5分間隔で経過を刻むので、完走すれば正確な所要が出る
2. その値が GHA 相場（1h56m〜3h39m）に収まるなら **g++ 不在は主因ではない**＝上記の他要因を当たる
3. 収まらないなら、`m2w64-toolchain`（conda）か MSVC Build Tools を入れて**同じパラメータで再測**し、差を記録する
4. 効果が無ければ「g++ は無関係だった」ことを**棄却として残す**（ADR-0024 の「棄却も含めて記録に残す」作法。書き残さないと「g++ を入れれば速くなるのでは」が何度も再提案される）

## 検証

- [ ] 導入前の所要（heartbeat の経過ログ・完走した実行で1本）
- [ ] 導入後の所要（同一パラメータ・同一データ世代）
- [ ] `pytensor.config.cxx` が空でなくなり、起動時の警告が消えること
- [ ] 差が無ければ棄却として記録してクローズ

## 関連

- #504（月次バッチのローカル移行。所要が伸びると「上限16時間」の前提に効く）
- PR #511（進捗可視化。これが無ければこの警告は今も見えていない）

**Comment by kome-kome** [comment]
## ローカル実測（進行中・2026-08-21）

PR #511 の heartbeat が入ったので、**初めて所要が観測できる状態**になった。現時点の実測:

| 項目 | 値 |
|---|---|
| 起動 | 2026-08-21 16:22:15 JST |
| 経過 | **345分（5時間45分）以上・継続中** |
| heartbeat | 69回（5分間隔） |
| パラメータ | `--draws 800 --tune 800`（`run_monthly --steps macro_beta`） |
| GHA 最速実績 | 116分（1時間56分） |
| 倍率 | **2.97倍以上**（まだ伸びている） |

本 Issue が立てた仮説（`pytensor: g++ not detected` → Python フォールバック）と整合する方向の数字だが、**これだけでは因果は確定しない**。draws/tune の値・パネル規模・マシン負荷が GHA 実行時と同一である保証がないため。

### 確定に必要なもの

- [ ] 完走後の総所要を記録（本コメントは進行中の値）
- [ ] 当時の GHA 実行ログからパネル規模（n_stock / n_sector / n_factor / n_obs）を取り、今回の「サンプリング開始」ログ（PR #511 で追加）と突き合わせる。**規模が違えば倍率の比較は無意味**
- [ ] g++ 導入後に同一パラメータ・同一パネルで再実行して比較

### 副産物

heartbeat そのものは意図どおり機能している。是正前は同じ実行が**無音のまま**で、「走っている」のか「死んだ」のかを区別できなかった（2026-08-21 の12時間誤認が発端）。現在は 5分ごとに経過が刻まれ、**遅いこと自体は分かる**状態になった。

heartbeat を `macro_beta_inference.py` 専用から `batch_common.Runner` へ移し全ステップへ効かせる件は #522 で別途起票した。

**Comment by kome-kome** [comment]
## 実測で切り分けた: 差の実体は「1 leapfrog 歩あたりのコスト」で、g++ ではない

本 Issue のコメントに書いた着手順（「どこに時間が溶けているかを先に測る」）に沿って計測した。計測ハーネスは `scripts/bench_macro_beta.py`（PR #533）、Linux 側の基準は `bench-macro-beta.yml` を dispatch して取った。

### 方法（比較が成立する形にするまでに3回作り直した）

- **決定的な合成パネル**（seed 固定）で両側同一データ・同一事後幾何にする。実データだと #503 の反転後は Supabase が 2026-08-07 断面でローカル正本と規模が違い、比が汚れる
- **総 leapfrog 歩数に対する回帰**で 1歩の実費を出す。draws を分母にすると run 間で steps/draw が変わった瞬間に傾きが汚染される
- **計測前に probe を1本流す**。JAX はコンパイル済みカーネルを使い回すため、probe 無しだと2点目が速くなり傾きが負に出る（実測: draws=20 が 15.1s、draws=60 が 7.7s ＝ 限界費 −0.186s/draw）

### 結果1: 本番規模の実データ（ローカル・chains=2・n_stock=3800 / n_obs=94,037）

| 項目 | 実測 |
|---|---|
| steps/draw | **1023.0 / 1020.1**（max treedepth 到達率 1.000 / 0.933） |
| 1歩あたり | **22,176 us/step**（= 0.2358 us/step/obs） |
| 限界費 | 45.18 s/draw |
| cpu/wall | **3.43〜3.67**（6コアのうち約3.5コア） |
| ピーク常駐 | **1,205MB**（draws=15 の時点） |
| フル規模への外挿 | **1,159分 ＝ 約20時間** |

外挿値は 8/21 の「741.5分で未完走・手動停止」と整合する（あのまま回していても終わっていない）。

**cpu/wall 3.5 は、8/21 のフル実行で観測した「3.86コアを継続使用」とほぼ一致する**＝今回の計測は当時と同じ状態を再現できている。

### 結果2: ローカル↔Linux ランナー（同一合成パネル・chains=1・tune=100）

| n_stock | n_obs | GHA (4 vCPU) us/step/obs | ローカル (i5-8400) us/step/obs | 比 |
|---|---|---|---|---|
| 125 | 3,000 | 0.0628 | 0.0914 / 0.0798 | 1.3〜1.5 |
| 250 | 6,000 | 0.0454 | 0.1911 / 0.1447（夜の静かな時間帯は 0.0594） | 1.3〜4.2 |
| 500 | 12,000 | 0.0357 | 0.2385 / 0.2928（同 0.1502） | 4.2〜8.2 |
| 1000 | 24,000 | **0.0356** | 0.1890 | 5.3 |
| 3800（実データ） | 94,037 | — | **0.2358** | **6.6**（GHA 頭打ち値との比） |

**向きが逆**なのが要点: GHA は規模が上がるほど観測1件あたりが**安く**なり 0.0356 で頭打ちになるのに、ローカルは**高く**なる。

### 結果3: 設定要因は主因ではない

n_stock=250 で条件を1つずつ変えた（いずれも steps/draw=1023・発散0）:

| 条件 | us/step/obs |
|---|---|
| 本番経路（chains=2・デバイス強制2・chain_method 既定） | 0.0682〜0.1166 |
| chains=1 | 0.0594〜0.1143 |
| chain_method=sequential | 0.0787〜0.0956 |
| chain_method=vectorized | 0.0655 |
| デバイス強制なし（実効100デバイス） | 0.1467 |
| スレッド上限 1 / 3 | 0.0869 / 0.1013 |

いずれもばらつきの範囲を出ない。**チェーン並列の効き方も両者同程度**（n=500 で 2チェーン化による1歩あたりの改善は ローカル 1.44倍・GHA 1.46倍）＝ `chain_method` もデバイス強制も主因ではない。

### g++ 仮説（本 Issue の起票時の見立て）は棄却方向

- 差が**規模依存**（n≤250 で 1.3〜1.5倍、n≥500 で 4〜6.6倍）。C バックエンド不在で Python フォールバックしているなら、規模に依らず一律に遅いはずで向きが合わない
- 時間が溶けているのは `pm.sample` の中（heartbeat が 740分の時点でまだ刻んでいた）で、そこは `nuts_sampler="numpyro"` ＝ JAX/XLA が回している
- **直接反証**の条件は用意した: `bench-macro-beta.yml` に `pytensor_cxx=off`（`PYTENSOR_FLAGS=cxx=`）を追加済み。Linux から C バックエンドだけ外して所要が変わらなければ棄却が確定する（マージ後に dispatch する）

### もう1つの律速: posterior の常駐サイズ

`beta` / `mu_sector` は `pm.Deterministic` ＝ **トレースに保存される**。保存量は 1 draw あたり約 25 x n_stock doubles で、本番（3800銘柄・draws 800・chains 2）だと:

- 25 x 3800 x 8B = **760KB / draw / chain**
- x 800 draws x 2 chains ≒ **1.2GB**（今回 draws=15 で観測した 1,205MB に、この 1.2GB が上乗せされる）

この PC の空き物理メモリは計測中 **1.8〜2.4GB** だった（Chrome / VSCode 等が常駐）。つまり本番実行は**空きメモリの境界に乗る**。BLAS の matmul ベンチ（compute-bound）は同時刻に 1.2倍しか遅くならないのに NUTS は 2.4倍遅くなる、という差もこれと整合する（メモリ帯域・常駐サイズに効く仕事）。

### 計測の作法として分かったこと（次に測る人向け）

1. **probe 無しの2点計測は成立しない**（コンパイル代が1点目にだけ乗る）
2. **tune を切り詰めるとサンプラーが別レジームへ落ちる**。tune=25 では steps/draw が 1023 → 63 まで動き発散が 78 出て、回帰の傾きが負になった。本番と同じ「max treedepth に張り付く」領域で測ること（ハーネス側に警告を実装済み）
3. **ローカルの値は時間帯で 2.4倍振れる**。同一設定・同一レジームでも 20:50 と 04:00 で 356 us/step と 870 us/step。規模比較は**同一セッション内で交互に測る**、条件比較は**反復して最速を採る**

### 環境指紋

```
ローカル : Windows 11 / i5-8400 (6C6T・2017) / RAM 7.9GB / cpu_count=6
           jax 0.10.2 / jaxlib 0.10.2 / numpyro 0.21.0 / pymc 5.28.5 / pytensor 2.38.3
           pytensor.cxx='' / floatX=float64 / jax_enable_x64=True
GHA      : ubuntu-latest / cpu_count=4 / 同一 pin
```

余談: `XLA_FLAGS` に `--xla_force_host_platform_device_count=100` が混ざるのは **pymc 自身**（`pymc/sampling/jax.py` が import 時に 100 を前置し、既存フラグを剥がす正規表現 `=.+\s` が末尾空白を要求するため剥がし損ねる）。実効デバイス数が 2 と 100 に割れるが、所要への寄与はばらつきの範囲だった。

### 次

原因の切り分けは付いたので、本 Issue は「何を選ぶか」の判断へ移る。9/1 の月次は #530 の予算で `exit=124` として落ちる（合意済みの既定）。

**Comment by kome-kome** [comment]
## 派生させた改善レバーと、残っている確認

計測から出た「効きうる手」を2つ分離した（本 Issue は原因の切り分けに留める）:

- **#540 全 draw が max treedepth に張り付いている**: `steps/draw` が 1023（=2^10−1）で到達率 1.000。ローカル・GHA の双方、合成・実データの双方で同じ。所要は `steps/draw × 1歩の実費` なので、**こちらはプラットフォームに依らず効く**（本 Issue の 6.6倍はマシン側の話）。ただし ESS/秒で測らないと「速いが役に立たないサンプラー」になるため、ハーネスへ ESS 計測を足すところから
- **#541 posterior 常駐 1.2GB**: `beta` / `mu_sector` が Deterministic ＝トレース保存。`beta` は保存済みの自由 RV から厳密に再構成できるので、確率構造を変えずに約 584MB 減らせる

## 残っている確認（PR #539 マージ後）

`bench-macro-beta.yml` へ `pytensor_cxx=off`（`PYTENSOR_FLAGS=cxx=`）を足してある。**Linux 側から C バックエンドだけ外して所要が変わらなければ、本 Issue の起票時の仮説（g++ 不在が主因）は棄却が確定する**。現状は「規模依存という向きが仮説と合わない」という状況証拠までで、直接の反証はまだ取っていない。

なお 9/1 の月次は #530 のステップ予算どおり `macro_beta` が `exit=124` で落ちる（合意済みの既定）。`tune×3` は走るので μ̂ の固着は解ける。

**Comment by kome-kome** [comment]
## g++ 仮説の直接反証（2026-08-25）— 棄却で確定

PR #539 のマージで `bench-macro-beta.yml` の `pytensor_cxx=off`（`PYTENSOR_FLAGS=cxx=`）が使えるようになったので、**Linux ランナーから C コンパイラだけ外して**同一パラメータ（合成パネル 500銘柄・chains=1・tune=100・numpyro/adapt_diag）で対を2組、同時刻に dispatch した。

| 組 | `cxx='/usr/bin/g++'` | `cxx=''`（＝ローカルと同条件） |
|---|---|---|
| 1組目 | 0.0394 us/step/obs | **0.0226** |
| 2組目 | 0.0410 us/step/obs | **0.0386** |
| 参考（既定＝あり・別 run） | 0.0357 | — |

いずれも steps/draw=1023・発散0 の同一レジーム。**C バックエンドを外しても遅くならない**（最速の観測はむしろ「なし」側）＝ 仮説が予測する「著しい劣化」は起きない。

機構としても整合する: 本番は `nuts_sampler="numpyro"` ＝ モデルは JAX へトランスパイルされ、**サンプリングのループに PyTensor の C バックエンドは通らない**。g++ がある側で `probe`（初回コンパイル込み）が 56.2s、無い側が 30.4s と短いのは、**C コンパイル代が起動時にだけ乗る**ことの現れで、1歩あたりのコストには効かない。

棄却は ADR-0002 へ記録した（PR #543）。書き残さないと「g++ を入れれば速くなるのでは」が再提案されるため（ADR-0024 の作法）。

## 本 Issue の答え

差の実体は **1 leapfrog 歩あたりのコストで、規模依存**。本番規模（3,800銘柄・n_obs 94,037・chains=2）でローカルは 22,176 us/step ＝ **フル規模換算 約20時間**で、8/21 の「741.5分で未完走」と整合する。設定要因（`chain_method` / ホストデバイス強制 / スレッド上限）はいずれもばらつきの範囲だった。

残したレバーは #540（`steps/draw` の 1023 張り付き＝プラットフォーム非依存）と #541（posterior 常駐 1.2GB）。9/1 の月次は #530 の予算どおり `macro_beta` が `exit=124` で落ち、`tune×3` は走る。

PR #543 のマージで本 Issue はクローズする。

**Comment by kome-kome** [comment]
## 「6.4倍遅い」がメモリ枯渇下の値である疑い（#584 へ）

2026-09-01 の月次バッチ初実走で、本 Issue の切り分けに直接効く実測が出た。

**マシンの資源（実走中の実測）**
- CPU: i5-8400（6コア6スレッド・E-core 無し）。負荷 **39%＝余っていた**
- 物理メモリ: **7.9GB / 空き 0.6GB**。プロセスの WorkingSetSize が 1974MB → 1013MB → 799MB と測るたび減少＝**ページアウト**

**同じマシン・同じメモリ状況でモデル間に 11倍の差**
- `tune:macro_risk_return`（M-1）: 16.7分/件・ツリー常駐 1,974MB
- `tune:macro_dlm`（M-3）: 1.5分/件・ツリー常駐 470MB

`scripts/bench_macro_beta.py` の `peak_rss_mb` は docstring に「**この PC の空きメモリは 2GB 前後**しかないので、1歩あたりのコストとは別に常駐サイズ自体が律速になりうる」と書いていた。仮説は立っていたのに、**`env_fingerprint()` が空き物理メモリを残していなかった**ため確定できず、「6.4倍遅い」という所要だけが残った（docstring は「CPU/メモリ」と言いながら実装にメモリが無かった）。

## 対応済み

- `env_fingerprint()` に `mem_total_mb` / `mem_avail_mb` を追加した。**所要を比べるときの前提条件**なので、これが無いと「ローカルは N 倍遅い」がマシンの話かコードの話か後から切り分けられない
- `peak_rss_mb` の ctypes 実装は `sysmem.py` へ切り出して委譲（`scripts/batch_common.py` の heartbeat も同じ源を使う）。**測るのはプロセスツリーの合計**——`venv\Scripts\python.exe` はランチャースタブで、`Popen` の pid 単体では実体 313MB が **4MB** に見える

## 次

#584 で M-1 / M-3 を並べて秒/件と peak RSS を実測する。`macro_beta` 固有の問題なのか、M-1 系（`macro_snapshots` の共有キャッシュを使う経路）全般なのかがそこで分かる。
